upip/upip.py

```diff
@@ -243,6 +249,113 @@ def cleanup():
         except OSError:
             print("Warning: Cannot delete " + fname)
 
+def parse_version(string):
+    """
+        Parse Version
+
+        This function takes a string and gets all versioning infromation for
+        pypi according to PEP508 and PEP440
+
+        Written by: Stephan Kashkarov 2019
+    """
+
+    pkg = {}
+    pkg['name'], versioning = re.match(name_regex, string).groups()
+    data = get_pkg_metadata(pkg['name'])
+
+    # Removes "Extra" arguments
+    if ";" in versioning:
+        versioning = versioning.split(";")[0].split(",")
+
+    # split version parameters ie "<=3.2.3, > 2.7.1, !=3.0.1" -> [('<=', [3, 2, 3]), ('>', [2, 7, 1]), ('!=', [3, 0, 1])]
+    # Does not support letters in the version name (i.e. 1.2.5rc3)
+    parameters = []
+    index = 0
+    while len(versioning):
+        index = versioning.find(',')
+        if index != -1:
+            substr = versioning[:index]
+            versioning = versioning[index + 1:]
+        else:
+            substr = versioning
+            versioning = ""
+        while substr[index - 1] not in ['=', '>', '<']:
+            index -= 1
+        operation = substr[:index].strip()
+        version = [int(x) if x != "*" else "*" for x in substr[index:].strip().split('.')]  # "3.0.1" -> [3, 0, 1] && "4.2.*" -> [4, 2, "*"]
+        parameters.append((operation, version))
+
+    versions = [[int(y) for y in x.split(".")] for x in data["releases"].keys()] # makes list of versions in ver_list format
+
+    # reduces list using operators
+    while parameters:
+        operator, version = parameters.pop(0)
+        if operator == "==":
+            # Check wildcard
+            if "*" in versions:
+                parameters.append((">=", version[:-1])) # queues wildcard
+                continue
+
+            # direct version
+            elif version in versions:
+                pkg['version'] = version
+                return pkg, data
+        
+        # removes wildcards
+        version.pop()
+
+        if operator == ">":
+            versions = filter(versions, lambda x: ver_list_cmp(version, x) < 0)
+
+        elif operator == ">=":
+            versions = filter(versions, lambda x: ver_list_cmp(version, x) <= 0)
+
+        elif operator == "<":
+            versions = filter(versions, lambda x: ver_list_cmp(version, x) > 0)
+
+        elif operator == "<=":
+            versions = filter(versions, lambda x: ver_list_cmp(version, x) >= 0)
+
+        # Arbitrary operator
+        elif operator == "~=":
+            parameters.extend([("<=", version[:-1]), (">=", version)])
+            continue
+
+    if not versions:
+        raise NoVersionError
+    # returns max suitable operator
+    pkg['version'] = max(versions)
+    return pkg, data
+
+def ver_list_cmp(list1, list2):
```

Comment:
Again, unit testing this would be relatively straightforward and beneficial!
Also, I believe that if you used tuples then the built-in tuple comparisons will work correctly. Haven't yet validated that though...